rfcs/text/0016_ols_phase_1.md

```diff
@@ -0,0 +1,323 @@
+
+### `delete`
+
+The security wrapper will first retrieve the requested `private` object to ensure the user is authorized.
+
+#### Performance considerations
+The security wrapper must first retrieve the existing `private` object to ensure that the user is authorized. This requires another round-trip to `get` the `private` object so we can authorize the operation.
+
+This overhead does not impact deleting "public" objects. We only need to retrieve objects that are registered as `private`. As such, we do not expect any meaningful performance hit initially, but this will grow over time as the feature is used.
+
+
+### `find`
+The security wrapper will supply or augment a [KQL `filter`](https://github.com/elastic/kibana/blob/701697cc4a34d07c0508c3bdf01dca6f9d40a636/src/core/server/saved_objects/types.ts#L118) which describes the objects the current user is authorized to see.
+
+```ts
+// Sample KQL filter
+const filterClauses = typesToFind.reduce((acc, type) => {
+  if (this.typeRegistry.isPrivate(type)) {
+    return [
+      ...acc,
+      // note: this relies on specific behavior of the SO service's `filter_utils`,
+      // which automatically wraps this in an `and` node to ensure the type is accounted for.
+      // we have added additional safeguards there, and functional tests will ensure that changes
+      // to this logic will not accidentally alter our authorization model.
+
+      // This is equivalent to writing the following, if this syntax was allowed by the SO `filter` option:
+      // esKuery.nodeTypes.function.buildNode('and', [
+      //   esKuery.nodeTypes.function.buildNode('is', `accessControl.owner`, this.getOwner()),
+      //   esKuery.nodeTypes.function.buildNode('is', `type`, type),
+      // ])
+      esKuery.nodeTypes.function.buildNode('is', `${type}.accessControl.owner`, this.getOwner()),
+    ];
+  }
+  return acc;
+}, []);
+
+const privateObjectsFilter =
+  filterClauses.length > 0 ? esKuery.nodeTypes.function.buildNode('or', filterClauses) : null;
+```
+
+#### Performance considerations
+We are sending a more complex query to Elasticsearch for any find request which requests a `private` saved object. This has the potential to hurt query performance, but at this point it hasn't been quantified.
+
+Since we are only requesting saved objects that the user is authorized to see, there is no additional overhead for Kibana once Elasticsearch has returned the results of the query.
+
+
+### `addToNamespaces` / `deleteFromNamespaces`
+
+The security wrapper will ensure that the user is authorized to share/unshare all existing `private` objects.
+#### Performance considerations
+Similar to the "create / override" scenario above, the security wrapper must first retrieve all of the existing `private` objects to ensure that the user is authorized. This requires another round-trip to `get`/`bulk-get` all `private` objects so we can authorize the operation.
```

Comment:
__question:__ 

> This requires another round-trip to `get`/`bulk-get` all `private` objects

If I read it correctly  `addToNamespaces` / `deleteFromNamespaces` work with single objects and we don't need to do a `bulk-get` here? Feel free to ignore this comment if we plan to extend these APIs to work with multiple objects in the future.

Reply:
This will be changing in the future to support bulk operations via a new `updateObjectsSpaces` function (https://github.com/jportner/kibana/blob/4d0b48e9828e27bc645a2a0b4883acc3cc4a3cc7/src/core/server/saved_objects/service/lib/repository.ts#L1271-L1271).

Otherwise, you're right that the current implementation does not support bulk operations

Reply:
Good to know, thanks!